### PR TITLE
Add command line globbing to all utilities (for windows targets)

### PR DESCRIPTION
**Pull request by rivy**
Command line globbing is added (for windows targets) to all utilities via 'uucore'.

This change increases portability and parity across machine targets.

This initial commit uses the published 'wild' crate. The second commit references a personal fork of 'wild' which updates the API and includes proper case insensitivity (again, for windows targets). I plan to update and re-reference the published 'wild' crate when it is updated to include the fixes.

### CODE REVIEW

**Comment by rivy** [comment]
It's not obvious how to fix the failing tests from my end.

I do believe they are due to bugs in `wild` and/or `glob`, but `glob` seems semi-deprecated and `wild` is hard for me to grok in order to issue a PR for repair of the problem (maybe the author of `wild` could offer a hand? @kornelski?).

Maybe the three failing tests should just be gated out for Windows for the time being until the globbing story becomes more polished (see rust-lang-nursery/glob#59).

**Comment by kornelski** [comment]
Note that Windows' quoting/escaping syntax is bizarre, and incompatible with any sensible Unix-like expectations. It has weird hacks to tolerate unescaped backslashes:

> * 2n backslashes followed by a quotation mark produce n backslashes followed by begin/end quote. This does not become part of the parsed argument, but toggles the "in quotes" mode.
> * (2n) + 1 backslashes followed by a quotation mark again produce n backslashes followed by a quotation mark literal ("). This does not toggle the "in quotes" mode.

So your test expectations are simply wrong for Windows.

**Comment by rivy** [comment]
Agreed, CMD has a LOT of weird edge cases. But the initial failing test that I see is `echo \\` returning just a single `\`. No quotes involved. In fact, when I tested the executable with various permutations of slashes, it will "compress" them.

```
C:> REM "compression" of slashes
C:> uutils echo \
\
C:> uutils echo \\
\
C:> uutils echo \\\\\\
\
C:> uutils echo \\.
\.
C:> uutils echo \\\\.
\.
C:> uutils echo /
/
C:> uutils echo ///////
/
C:> uutils echo //.
/.
C:> uutils echo ////.
/.
C:> REM but if not a possible PATH? ...
C:> uutils echo \\x //x
\\x //x
```

I think the wildcard/globbing process is being overly aggressive. If there aren't any glob characters in an argument (or of globbing returns an empty set), the argument should just be returned. Isn't that a wild/globbing issue?

**Comment by kornelski** [comment]
Fixed in wild v1.1.1 and v2.0.1

**Comment by rivy** [comment]
👍 

Looking better, and passing all tests, on my machine... Let's see how we do with AppVeyor.

**Comment by rivy** [comment]
Excellent, thanks @kornelski !

File: src/uucore/lib.rs [hunk 1]
```diff
@@ -1,13 +1,9 @@
 extern crate wild;
 
-pub fn args() -> Box<Iterator<Item=String>> {
+pub fn args() -> impl Iterator<Item=String> {
```
Comment (vks) [inline src/uucore/lib.rs]: Is this supported by Rust 1.22?

**Comment by vks** [comment]
This is problematic, because `*` is a valid file or folder name on Unix. If I understand correctly, this PR makes it impossible to remove such files and folders.

Only enabling the feature on Windows might work, because there `*` is not allowed in file and directory names.

**Comment by kornelski** [comment]
The `wild` crate is a no-op on everything but Windows

File: src/uucore/lib.rs (hunk 1 above)
Comment (rivy) [reply src/uucore/lib.rs]: No, `impl Trait` support didn't land until 1.26 (see <https://blog.rust-lang.org/2018/05/10/Rust-1.26.html>). I'm happy to revert it to using boxing. It does compile, test, and run correctly with the boxing version, though it's likely slightly (probably undetectably) less performant.

But, FYI, in testing this, I can't get the repo to compile using the "1.22.1" toolchain anyway. There are lots of "f32::from_bits(...)" and similar errors.

To make the backwards compatibility more explicit, it might be a good idea to modify the AppVeyor CI to test back to the minimum version that you'd like to support. There's an example configuration being used by [globwalk](https://github.com/rivy/rust.globwalk/blob/master/appveyor.yml) that I could modify into a PR which tests more widely if that's a help.

Comment (Arcterus) [reply src/uucore/lib.rs]: I've fine with bumping up to 1.26.  I want to use `impl Trait` too (lol).  If you could modify that in `README.md` I'll merge this as is.

Comment (rivy) [reply src/uucore/lib.rs]: Sure, I'll be happy to update the `README.md`. While I'm doing that, I'll also double-check that it actually compiles with 1.26 as well.

Comment (rivy) [reply src/uucore/lib.rs]: Unfortunately, rust@1.26 isn't enough. It looks like rust@1.27 is the actual minimum needed.

Several external crates, which were upgraded to newer versions in prior commits, are using `from_bits` which causes compile-time errors. I can't find a definitive source (eg, CHANGELOG) that mentions when `from_bits` was landed into stable mainline, but I've found a discussion that says rust@1.27 is needed (see <https://github.com/starkat99/half-rs/issues/11#issuecomment-404700575>). And I found that to be the case, as well.

Compilation and testing work fine with rust@1.27 (and I added a specific test to the Travis CI configuration for the minimum rust version). There is a Travis CI caveat (see below). Would you like me to change the minimum to 1.27? The alternative is to revert all the external crate upgrades which are requiring 1.27.

Notably, Travis CI builds are failing because of use of "Sphinx" 1.8.0 during install testing. Older builds succeeded using Sphinx@1.7.8. I have no experience with that build process, and I don't have a suggested solution (other than to cut it out).

There's also a bit of old code cruft that can be taken out to avoid all of the build warnings.

Comment (rivy) [reply src/uucore/lib.rs]: @Arcterus , I've got a branch with rust@1.27 and revised / working Travis CI testing at <https://github.com/rivy/rust.coreutils/tree/fix.ci>, specifically <https://github.com/rivy/rust.coreutils/commit/deead6f672925cab72e319770f9214b1cd07741a>.

AppVeyor CI was failing for odd reasons, so I remodeled it based upon a newer template (see <https://github.com/starkat99/appveyor-rust/blob/master/appveyor.yml>). 

The Travis CI build is at <https://travis-ci.org/rivy/rust.coreutils/builds/432010513>.

The AppVeyor CI build is at <https://ci.appveyor.com/project/rivy/rust-coreutils/build/2%20~%20fix.ci>.

I can add all or part of it to this PR... let me know how you'd like me to proceed.

**Comment by rivy** [comment]
@Arcterus , I've got a branch with rust@1.27 and revised / working Travis CI testing at <https://github.com/rivy/rust.coreutils/tree/fix.ci>, specifically <https://github.com/rivy/rust.coreutils/commit/deead6f672925cab72e319770f9214b1cd07741a>.

AppVeyor CI was failing for odd reasons, so I remodeled it based upon a newer template (see <https://github.com/starkat99/appveyor-rust/blob/master/appveyor.yml>). 

The Travis CI build is at <https://travis-ci.org/rivy/rust.coreutils/builds/432010513>.

The AppVeyor CI build is at <https://ci.appveyor.com/project/rivy/rust-coreutils/build/2%20~%20fix.ci>.

I can add all or part of it to this PR... let me know how you'd like me to proceed.

((  reposted from sub-discussion above ))

**Comment by Arcterus** [comment]
It’s fine if you include it in this PR.

**Comment by rivy** [comment]
I've used this updated PR as a base for further fixes, and it seems like a good new base.

All tests are working correctly for all of my newer commits (which I'll rebase and PR after this is merged).

**Comment by cnd** [comment]
@Arcterus can we merge this?

**Comment by rivy** [comment]
@Arcterus ,

Just FYI, I've got several pending PRs based on this now (#1304, #1305, #1315, #1316, #1317, #1318).

**Comment by rivy** [comment]
@Arcterus @cnd 

I just wanted to check in now that the holidays are over and see if we can get this merged (as well as the other PR's mentioned above) ...

**Comment by rivy** [comment]
@vks ?

**Comment by Arcterus** [comment]
Really sorry about being MIA for the past while.  This seems good to me though!
